table/block_based_table_builder.cc

```diff
@@ -659,15 +681,50 @@ static void DeleteCachedBlockContents(const Slice& /*key*/, void* value) {
   delete bc;
 }
 
+template <class Entry>
+void DeleteCachedEntry(const Slice& /*key*/, void* value) {
+  auto entry = reinterpret_cast<Entry*>(value);
+  delete entry;
+}
+
 //
 // Make a copy of the block contents and insert into compressed block cache
 //
 Status BlockBasedTableBuilder::InsertBlockInCache(const Slice& block_contents,
                                                   const CompressionType type,
-                                                  const BlockHandle* handle) {
+                                                  const BlockHandle* handle,
+                                                  bool is_data_block) {
   Rep* r = rep_;
-  Cache* block_cache_compressed = r->table_options.block_cache_compressed.get();
 
+  // Uncompressed regular block cache
+  if (r->table_options.prepopulate_data_blocks && is_data_block) {
+    assert(r->level == 0);
+    Cache* block_cache = r->table_options.block_cache.get();
+    if (type == kNoCompression && block_cache != nullptr) {
+      size_t size = block_contents.size();
+      auto buf = AllocateBlock(size, block_cache->memory_allocator());
+      memcpy(buf.get(), block_contents.data(), size);
+      BlockContents results(std::move(buf), size);
+
+      char cache_key[BlockBasedTable::kMaxCacheKeyPrefixSize + kMaxVarint64Length];
+      Slice key = BlockBasedTable::GetCacheKey(
+          r->cache_key_prefix, r->cache_key_prefix_size, *handle, cache_key);
+
+      Block* block =
+          new Block(std::move(results), kDisableGlobalSequenceNumber);
+      size_t charge = block->ApproximateMemoryUsage();
+
+      block_cache->Insert(key, block, charge, &DeleteCachedEntry<Block>);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_ADD);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_DATA_ADD);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_BYTES_WRITE, charge);
+      RecordTick(r->ioptions.statistics, BLOCK_CACHE_DATA_BYTES_INSERT, charge);
+      return Status::OK();
+    }
+  }
```

Comment:
I'm confused a little bit. It feels that compress and uncompress cache are totally separated logic. Can we make them in two different functions, and call them accordingly?
I feel it is clearer than a if here and return in the end of the if. The branching can happen in the caller.